Collaboration Center | Dev | Tabs component

danielamorse opened the pull request:
## Jira

http://vjira2:8080/browse/BDS-1437

## Summary

Add Tabs component.

## Details

This is the first pass at the Tabs component. It should have all the features needed for Collaboration Center. Additional features, such as `variant`, will be handled separately.

## How to test
### Testing steps
- Review new files, mainly:
    - [tabs.twig](https://github.com/bolt-design-system/bolt/pull/1337/files#diff-cd51b2cf00f0c5fe702b9279b2ba6c10)
    - [tabs.scss](https://github.com/bolt-design-system/bolt/pull/1337/files#diff-d03ab74025ee67bc053491da54641be0)
    - [tabs.js](https://github.com/bolt-design-system/bolt/pull/1337/files#diff-efce6a3a9673d446e153d19f4862684d)
    - [TabPanel/tab-panel.twig](https://github.com/bolt-design-system/bolt/pull/1337/files#diff-94adabe023a4fbc71ff4eee18744165f)
    - [TabPanel/tab-panel.scss](https://github.com/bolt-design-system/bolt/pull/1337/files#diff-b0497c731b98dd96157b4d5b7213236e)
    - [TabPanel/index.js](https://github.com/bolt-design-system/bolt/pull/1337/files#diff-55103e1cb75fba8e6f188f572d828905)
- Review all demos with and without JS.
- Test in IE11.

### Caveats
- ~~`align` prop doesn't work when JS is off because the underlying markup cannot support it.~~ Works now!
- The labels are not navigable by tabbing when JS is off, and I'm not sure why. Follow up ticket or can I get some help with this one?
- There are a bunch of JS warnings that read `$ref: keywords ignored in schema at path`
    - Sounds like same issue as: https://github.com/cloudevents/sdk-javascript/issues/7
    - Is this actually a problem? A result of the version of AJV we're using?
- There is a [separate ticket](http://vjira2:8080/browse/BDS-1787) for adding more tests to this component. I don't want that to block review of the component itself.

sghoweri commented on docs-site/src/pages/pattern-lab/_patterns/02-components/tabs/05-tabs.twig:
```diff
@@ -0,0 +1,17 @@
+  items: [
+    {
+      label: "Tab label 1",
+      content: "<p>Tab content #1. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent porta ut tortor id luctus. Cras fermentum pretium ipsum sit amet tempus. Sed lobortis dui ante, feugiat pharetra turpis dapibus sit amet. Donec hendrerit diam vel ante hendrerit ornare. Donec non aliquam nisi. Praesent malesuada mollis dui quis imperdiet. Donec dictum quam purus, ut viverra est gr ...
```
@danielamorse out of curiosity, how does the height of this behave / transition if, for example, the first tab has 1 line of text vs the other tabs have 20+ lines of text?

danielamorse replied:
@sghoweri the Tab container height is determined by the height of the visible tab panel. So, it will jump up and down if content height changes. Perhaps fixed height would be a nice add-on.

danielamorse replied:
@sghoweri @mikemai2awesome, added this ticket for future Tabs height options: http://vjira2:8080/browse/BDS-1816

sghoweri commented on docs-site/src/pages/pattern-lab/_patterns/02-components/tabs/10-tabs-spacing-variations.twig:
```diff
@@ -0,0 +1,26 @@
+{% set schema = bolt.data.components["@bolt-components-tabs"].schema %}
+
+<p>
+  By setting the <code>spacing</code> prop to <code>none</code>, <code>xsmall</code>, <code>small</code>, <code>medium</code>, or <code>large</code>, it applies the appropriate inset spacing for each tab panel and label. The default is <code>medium</code>.
```
Just a friendly reminder to everybody (not trying to single you out @danielamorse  😉): these are the types of docs that’ll go out of sync if we aren’t really really careful.

Perhaps the solution is for us to use MDX or some other pseudo-dynamic doc format... or maybe even just a few low level Twig macros...

Either way, if we ever change our component prop options or names, these hard coded doc references tend to be brittle!

danielamorse replied:
@sghoweri agreed, removed! It's all there in the schema anyway.

sghoweri commented on docs-site/src/pages/pattern-lab/_patterns/02-components/tabs/10-tabs-spacing-variations.twig:
```diff
@@ -0,0 +1,26 @@
+    items: [
+      {
+        label: "Tab label 1",
+        content: "<p>Tab content #1. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent porta ut tortor id luctus. Cras fermentum pretium ipsum sit amet tempus. Sed lobortis dui ante, feugiat pharetra turpis dapibus sit amet. Donec hendrerit diam vel ante hendrerit ornare. Donec non aliquam nisi. Praesent malesuada mollis dui quis imperdiet. Donec dictum quam purus, ut viverra est  ...
```
Did you know that we have a lorum-ipsum generator in Bolt Core PHP?

danielamorse replied:
@sghoweri I like that idea a lot. Do you think it will be a problem for the VRT testing done by QA or for our own Nightwatch tests if it's going to be randomly generated every time?

sghoweri commented on docs-site/src/pages/pattern-lab/_patterns/02-components/tabs/999-tabs-web-component.twig:
```diff
@@ -0,0 +1,133 @@
+{% set basic_usage %}
+<bolt-tabs>
+  <bolt-tab-panel>
+    <bolt-text slot="label">Tabs item 1</bolt-text>
```
What’s the resulting CSS styles for the text in these labels? 

Is it the custom bolt-text styles or is it the default tab label styles?

danielamorse replied:
@sghoweri it's the default tab label styles. In that case, it would be better to just use a `<div>` I think. I'll update.

sghoweri commented on docs-site/src/pages/pattern-lab/_patterns/02-components/tabs/999-tabs-web-component.twig:
```diff
@@ -0,0 +1,133 @@
+{% macro code_example(code) %}
```
Meta comment: my gut tells me that these WC demos aren’t easily maintainable + could get heavily templated.

sghoweri commented on packages/components/bolt-tabs/CHANGELOG.md:
```diff
@@ -0,0 +1,16 @@
+# Change Log
```
Wait a sec — how’s there already a CHANGELOG file?

danielamorse replied:
@sghoweri by duplicating an existing component to start, whoops! Removed.

sghoweri commented on docs-site/src/pages/pattern-lab/_patterns/02-components/tabs/00-tabs-docs.twig:
```diff
@@ -0,0 +1,23 @@
+{% set usage %}{% verbatim %}
+{% include "@bolt-components-tabs/tabs.twig" with {
+  items: [
```
`items` vs `panels`?

danielamorse replied:
@sghoweri I like `panels` better. Going with that, thanks 👍

mikemai2awesome commented on docs-site/src/pages/pattern-lab/_patterns/02-components/tabs/25-tabs-theme-variations.twig:
```diff
@@ -0,0 +1,28 @@
+            panels: [
+              {
+                label: "Tab label 1",
+                content: "<p>Tab panel 1. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent porta ut tortor id luctus. Cras fermentum pretium ipsum sit amet tempus. Sed lobortis dui ante, feugiat pharetra turpis dapibus sit amet. Donec hendrerit diam vel ante hendrerit ornare. Donec non aliquam nisi.</p>",
```
Please remove any naked elements, we don't want to encourage people to do that. For the demo code, use all twig. For the doc ui, differentiate by using web components.

danielamorse replied:
Fixed 👍

danielamorse commented:
@remydenton, thanks for looping in the rest of the integration crew. I want to be sure we deliver something that works for you.

Just want to call your attention to [this part](https://github.com/bolt-design-system/bolt/pull/1337/files#diff-cd51b2cf00f0c5fe702b9279b2ba6c10R47), where we're passing data from `tabs.twig` into each `tab-panel.twig` template. I tried to avoid this but it was the only way (in Twig) to get the API we wanted, i.e. we want to set shared props like `spacing` and `inset` on the parent, not on each child. That works great on web components, but I'm not sure about Drupal. I take it you saw this, and it's something you can handle on the Drupal-side?

remydenton commented:
Yep, that's exactly what I was looking at @danielamorse.  While the ideal is to effectively have a separate twig template for each custom element, I see what you're doing and decided it wouldn't be _that_ much of a problem on the Drupal side.

To put it in Drupal terms: if you had a view of entities, and you wanted to display the entities within tabs, it's not that critical to be able to theme the views list and each views item wrapper separately.  Instead, you could just find the views list template where you could access an array of entity render arrays, then reformat that array so that you could pass it along to `tabs.twig`.  The most important part is that the `content` for each tab could still be a render array representing each item in the list.

TLDR: I initially thought it was a big deal, but now I've decided it's a small deal and it's fine as-is.  Can't wait for proper SSR though so we can just use web components instead of twig!

charginghawk commented:
Agree with Rémy that this is good. Just to hit the nail on the head, the open-ended-ness of `content` is really important for us, because Drupal renders children first and pulls them into the parent templates. So, if I have a teaser, and in the teaser is a button, Drupal wants to render the button, and just dump that markup into the teaser. However, if you require the button schema in the teaser template (which is currently the case), we have to do extra work to flip around the data (and of course, we do, but it can be brittle and counter-intuitive for Drupal developers).

danielamorse commented:
@mikemai2awesome I updated the active label color ([see commit](https://github.com/bolt-design-system/bolt/pull/1337/commits/44c91c3b4e015faf46e4fb0ebb0c28d83f658f4c)) and changed cursor to always `pointer` ([see commit](https://github.com/bolt-design-system/bolt/pull/1337/commits/32bcecc80eb36c3e59e1b71c5cd347a0640f66e7)). Good call.

On a related note, I updated the `tabindex` behavior to match the specs on [w3.org](https://www.w3.org/TR/wai-aria-practices/examples/tabs/tabs-1/tabs.html), i.e. tab key places you on the active tab and skips all inactive tabs. Use arrow keys to switch between tabs.

sghoweri commented:
Taking a look at this tomorrow — 2nd thing on my todo list once I get in 👍🏼
